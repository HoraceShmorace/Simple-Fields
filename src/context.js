import { createContext, useReducer } from 'react'
import { node } from 'prop-types'
const ADD_DATA_ACTION = 'ADD_DATA'
const RESET_DATA_ACTION = 'RESET_DATA'
const initialState = {}
const reducer = (state, { payload, type }) => {
  switch (type) {
    case ADD_DATA_ACTION:
      return {
        ...state,
        ...payload
      }
    case RESET_DATA_ACTION:
      return {}
    default:
      return state
  }
}

export const Context = createContext(initialState)

export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)
  const addData = payload => dispatch({ payload, type: ADD_DATA_ACTION })
  const resetData = () => dispatch({ payload: initialState, type: RESET_DATA_ACTION })

  return (
    <Context.Provider value={{ addData, resetData, state }}>
      {children}
    </Context.Provider>
  )
}

ContextProvider.propTypes = {
  children: node.isRequired
}
