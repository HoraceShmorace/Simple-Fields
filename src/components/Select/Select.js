import { useContext, useEffect } from 'react'
import { array, string } from 'prop-types'
import { Context } from '~src/context'

const Select = ({
  default: defaultValue,
  id,
  options,
  required,
  value: initialValue
}) => {
  const dataId = `field_${id}`
  const isRequired = required === '1'
  const { addData, state } = useContext(Context)
  const onChange = e => addData({ [dataId]: e.target.value })

  useEffect(() => {
    const value = initialValue || defaultValue
    addData({ [dataId]: value })
  }, [defaultValue, initialValue])

  return (
    <select
      aria-required={isRequired}
      id={id}
      name={id}
      onChange={onChange}
      required={isRequired}
      value={state[dataId]}
    >{
      options.map(({ label, value }) =>
        <option
          key={value}
          value={value}
        >
          {label}
        </option>
      )
    }
    </select>
  )
}

Select.propTypes = {
  id: string.isRequired,
  options: array,
  value: string
}

Select.defaultProps = {
  options: [],
  value: 'Option2'
}

export default Select
