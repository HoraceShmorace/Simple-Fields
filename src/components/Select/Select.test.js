import { render } from 'enzyme'
import toJson from 'enzyme-to-json'
import Select from './Select'
import { ContextProvider } from '~src/context'
import mockData from '~mock/form.json'

const FIELD_TYPE_TEXT = 'select'
const { fields, id } = mockData

describe('Select', () => {
  describe('when rendered', () => {
    const props = fields.find(({ type }) => type === FIELD_TYPE_TEXT)
    const component = render(
      <ContextProvider formId={id}>
        <Select {...props} />
      </ContextProvider>
    )

    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot()
    })
  })
})
