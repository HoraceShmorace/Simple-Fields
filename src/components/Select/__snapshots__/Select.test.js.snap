// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select when rendered should match snapshot 1`] = `
<select
  aria-required="false"
  id="115390365"
  name="115390365"
>
  <option
    value="Option1"
  >
    Option1
  </option>
  <option
    value="Option2"
  >
    Option2
  </option>
  <option
    value="Option3"
  >
    Option3
  </option>
</select>
`;
